source/fe/fe_enriched.cc

```diff
@@ -0,0 +1,949 @@
+    for (unsigned int fe=1; fe < fes.size(); fe++)
+      {
+        if (dynamic_cast<const FE_Nothing<dim>*>(fes[fe]) == 0) // this is not FE_Nothing
+          {
+            // that's going to be an enriched element
+            Assert ( multiplicities[fe] == functions[fe-1].size(),
+                     ExcMessage("fnctions["+std::to_string(fe-1)+"] is expected to be of size "+std::to_string(multiplicities[fe])));
+
+            enriched = true;
+          }
+      }
+    return enriched;
+  }
+
+
+}
+
+
+template <int dim, int spacedim>
+const std::vector<std::vector<std::function<const Function<dim> *(const typename Triangulation<dim, spacedim>::cell_iterator &) > > >
+FE_Enriched<dim,spacedim>::get_enrichments() const
+{
+  return enrichments;
+}
+
+
+template <int dim, int spacedim>
+double
+FE_Enriched<dim,spacedim>::shape_value(const unsigned int   i,
+                                       const Point< dim > &p) const
+{
+  Assert(!is_enriched, ExcMessage("For enriched finite elements shape_value() can not be defined on the reference element."));
+  return fe_system.shape_value(i,p);
+}
+
+
+template <int dim, int spacedim>
+FiniteElement<dim,spacedim> *
+FE_Enriched<dim,spacedim>::clone() const
+{
+  std::vector< const FiniteElement< dim, spacedim > * > fes;
+  std::vector< unsigned int > multiplicities;
+
+  for (unsigned int i=0; i<this->n_base_elements(); i++)
+    {
+      fes.push_back( & base_element(i) );
+      multiplicities.push_back(this->element_multiplicity(i) );
+    }
+
+  return new FE_Enriched<dim,spacedim>(fes, multiplicities, get_enrichments());
+}
+
+
+template <int dim, int spacedim>
+UpdateFlags
+FE_Enriched<dim,spacedim>::requires_update_flags (const UpdateFlags flags) const
+{
+  UpdateFlags out = fe_system.requires_update_flags(flags);
+
+  if (is_enriched)
+    {
+      // if we ask for values or gradients, then we would need quadrature points
+      if (flags & (update_values | update_gradients))
+        out |= update_quadrature_points;
+
+      // if need gradients, add update_values due to product rule
+      if (out & update_gradients)
+        out |= update_values;
+    }
+
+  return out;
+}
+
+
+template <int dim, int spacedim>
+template <int dim_1>
+typename FiniteElement<dim,spacedim>::InternalDataBase *
+FE_Enriched<dim,spacedim>::setup_data (typename FiniteElement<dim,spacedim>::InternalDataBase *fes_data,
+                                       const UpdateFlags      flags,
+                                       const Mapping<dim,spacedim>    &mapping,
+                                       const Quadrature<dim_1> &quadrature,
+                                       internal::FEValues::FiniteElementRelatedData< dim, spacedim >   &output_data,
+                                       const bool subface) const
+{
+  typename FESystem<dim,spacedim>::InternalData *data_fesystem =
+    static_cast<typename FESystem<dim,spacedim>::InternalData *> (fes_data);
+  Assert (data_fesystem != NULL,
```

Comment:
i guess i can stick with `static_cast` and remove `Assert` because this function is called only inside the class always with `fe_system.get_face_data()` and alike.